tiny_dnn/layers/layer.h

```diff
@@ -404,48 +404,94 @@ class layer : public node {
    * This can be used during training phase when parameter update is
    * required.
    *
-   * @param trainable_only flag to return only the trainable parameters.
+   * @param trainable_only flag to return only the trainable parameters
    * @return std::vector of pointers to parameters
    */
   Parameters parameters(bool trainable_only = false) {
     Parameters parameters;
     for (size_t i = 0; i < parameters_.size(); i++) {
       if (!trainable_only ||
-          (ith_parameter(i).is_trainable() && trainable_only)) {
-        parameters.push_back(&ith_parameter(i));
+          (parameter_at(i).is_trainable() && trainable_only)) {
+        parameters.push_back(&parameter_at(i));
       }
     }
     return parameters;
   }
 
-  /**
-   * @brief Get const pointers to parameters of this layer.
-   *
-   * This can be used during inference phase when parameter update is not
-   * required.
-   *
-   * @param trainable_only flag to return only the trainable parameters.
-   * @return const std::vector of const pointers to parameters
-   */
-  const ConstParameters parameters(bool trainable_only = false) const {
+  ConstParameters parameters(bool trainable_only = false) const {
     ConstParameters parameters;
     for (size_t i = 0; i < parameters_.size(); i++) {
       if (!trainable_only ||
-          (ith_parameter(i).is_trainable() && trainable_only)) {
-        parameters.push_back(&ith_parameter(i));
+          (parameter_at(i).is_trainable() && trainable_only)) {
+        parameters.push_back(&parameter_at(i));
       }
     }
     return parameters;
   }
 
-  Parameter &ith_parameter(size_t i) { return *parameters_[i]; }
+  /**
+   * @brief Get pointers to weights of this layer.
+   *
+   * @param trainable_only flag to return only the trainable weights
+   * @return std::vector of pointers to parameters which are weights
+   */
+  // todo (karandesai) : rename after old weights method is removed
+  Parameters weights_p(bool trainable_only = false) {
```

Comment:
I see a lot of duplicated code, cool solution for this
https://stackoverflow.com/questions/9034388/c-how-to-make-find-if-not-only-return-the-first-object-for-which-applying-pred?answertab=votes#tab-top